src/sys/unix.rs

```diff
@@ -659,6 +665,75 @@ impl SockAddr {
+                    // Abstract addresses only exist on Linux.
+                    || (cfg!(not(any(target_os = "linux", target_os = "android")))
+                    && storage.sun_path[0] == 0)
+            })
+            .unwrap_or_default()
+    }
+
+    /// Returns the underlying `sockaddr_un` object if this addres is from the `AF_UNIX` family,
+    /// otherwise returns `None`.
+    pub(crate) fn as_sockaddr_un(&self) -> Option<&libc::sockaddr_un> {
+        self.is_unix().then(|| {
+            // SAFETY: if local, i.e. the `ss_family` field is `AF_UNIX` then storage must be a
+            // `sockaddr_un`.
+            unsafe { &*self.as_ptr().cast::<libc::sockaddr_un>() }
+        })
+    }
+
+    /// Get the length of the path bytes of the address, not including any terminating null.
+    fn path_len(&self, storage: &libc::sockaddr_un, null_terminated: bool) -> usize {
+        self.len() as usize - offset_of_path(storage) - if null_terminated { 1 } else { 0 }
+    }
+
+    /// Get a u8 slice for the bytes of the pathname or abstract name.
+    fn path_bytes(&self, storage: &libc::sockaddr_un, null_terminated: bool) -> &[u8] {
+        let path_len = self.path_len(storage, null_terminated);
+        // SAFETY: the pointed objects of type `i8` have the same memory layout as `u8`. The path is
+        // the last field in the storage and so it length is equal to
+        //          TOTAL_LENGTH - OFFSET_OF_PATH -1        if the path is null-terminated.
+        //          TOTAL_LENGTH - OFFSET_OF_PATH           if the path is not null-terminated.
+        // There is no safe way to convert a `&[i8]` ot `&[u8]`
+        unsafe { slice::from_raw_parts(storage.sun_path.as_ptr() as *const u8, path_len) }
+    }
+
+    /// Returns this address as a `Path` if it is an `AF_UNIX` pathname address, otherwise returns
+    /// `None`.
+    pub fn as_pathname(&self) -> Option<&Path> {
+        self.as_sockaddr_un().and_then(|storage| {
+            (self.len() > offset_of_path(storage) as u32 && storage.sun_path[0] != 0).then(|| {
+                // The -1 is for the terminating null.
+                let path_slice = self.path_bytes(storage, true);
+                Path::new::<OsStr>(OsStrExt::from_bytes(path_slice))
+            })
+        })
+    }
+
+    /// Returns this address as a slice of bytes representing an abstract address if it is an
+    /// `AF_UNIX` abstract address, otherwise returns `None`.
+    ///
+    /// Abstract addresses are a Linux extension, so this method returns None on all non-Linux
+    /// platforms.
+    pub fn as_abstract_namespace(&self) -> Option<&[u8]> {
+        #[cfg(any(target_os = "linux", target_os = "android"))]
+        {
+            self.as_sockaddr_un().and_then(|storage| {
+                (self.len() > offset_of_path(storage) as u32 && storage.sun_path[0] == 0)
+                    .then(|| self.path_bytes(storage, false))
```

Comment:
I think this should skip the first byte.

Reply:
Right, thank you. Changed inside `path_bytes`.